File: uniprot/writeUniprotProjectXML.pl
#!/usr/bin/perl

use strict;
use warnings;

use Getopt::Std;
use XML::LibXML;

use lib '../modules/perl';
require ImProjectXml;

use feature ':5.12';

my $usage = "Usage:writeUniprotProjectXML.pl [-h] [-i <project-xml-path>] <data-path> <taxon-id-path>

synopsis: read the taxon id file and creates Uniprot InterMine project XML <source> entry
If a <project-xml-path> is given then this entry is inserted directly into the given project XML
otherwise it is printed on stdout

options:
\t-h\tthis usage
";

my (%opts, $insert);

getopts('hi:', \%opts);
defined $opts{"h"} and die $usage;
defined $opts{"i"} and $insert++;
my $insertPath = $opts{i};

@ARGV > 1 or die $usage;

my ($dataPath, $taxonIdPath) = @ARGV;

open TAXONS, $taxonIdPath or die "Could not open $taxonIdPath: $!";
my $taxonIds = <TAXONS>;
chomp($taxonIds);
close TAXONS;

{
  my $sourceXml 
    = ImProjectXml::generateSource(
      "uniprot", "uniprot", 1,
      [
        [ 'uniprot.organisms', $taxonIds ],
        [ 'createinterpro', 'true' ],
        [ 'creatego', 'true' ],
        [ 'src.data.dir', 'location', $dataPath ]
      ]);

  if ($insert) {
    ImProjectXml::insertSourceIntoProjectXml($insertPath, $sourceXml);
  } else {
    say $sourceXml;
  }
}

{
  my $sourceXml
    = ImProjectXml::generateSource(
      "uniprot-fasta", "fasta", 0,
      [
        [ 'fasta.taxonId', $taxonIds ],
        [ 'fasta.className', 'org.intermine.model.bio.Protein' ],
        [ 'fasta.classAttribute', 'primaryAccession' ],
        [ 'fasta.dataSetTitle', 'UniProt data set' ],
        [ 'fasta.dataSourceName', 'UniProt' ],
        [ 'src.data.dir', 'location', $dataPath ],
        [ 'fasta.includes', 'uniprot_sprot_varsplic.fasta' ],
        [ 'fasta.sequenceType', 'protein' ],
        [ 'fasta.loaderClassName', 'org.intermine.bio.dataconversion.UniProtFastaLoaderTask' ]
      ]);

  if ($insert) {
    ImProjectXml::insertSourceIntoProjectXml($insertPath, $sourceXml);
  } else {
    say $sourceXml;
  }
}

{
  my $sourceXml
    = ImProjectXml::generateSource(
      "uniprot-keywords", "uniprot-keywords", 1,
      [
        [ 'src.data.dir', 'location', $dataPath ],
        [ 'src.data.dir.includes', 'keywlist.xml' ]
      ]);

  if ($insert) {
    ImProjectXml::insertSourceIntoProjectXml($insertPath, $sourceXml);
  } else {
    say $sourceXml;
  }
}

File: uniprot/writeUniprotProjectXml.py
#!/usr/bin/env python

import os
import sys

sys.path.insert(1, os.path.dirname(os.path.abspath(__file__)) + '/../modules/python')
import intermine.project as imp
import intermine.utils as imu
import synbio.dataset as sbds

############
### MAIN ###
############
parser = imu.ArgParser('Add Uniprot source entries to InterMine SynBioMine project XML.')
parser.add_argument('datasetPath', help='path to the dataset location.')
parser.add_argument('-v', '--verbose', action="store_true", help="verbose output")
args = parser.parse_args()

datasetPath = args.datasetPath
ds = sbds.Dataset(datasetPath)
uniprotDataPath = 'data/current/uniprot'

logPath = "%s/logs/writeOrthoDbProjectXml.log" % datasetPath
sys.stdout = imu.Logger(logPath)
projectXmlPath = "%s/intermine/project.xml" % datasetPath
taxonsPath = "%s/taxons/taxons.txt" % datasetPath

with open(taxonsPath) as f:
    taxons = f.read().strip()

imp.addSourcesToProject(
    "%s/intermine/project.xml" % datasetPath,
    [
        imp.Source('uniprot', 'uniprot',
        [
            { 'name':'src.data.dir',        'location':uniprotDataPath },
            { 'name':'uniprot.organisms',   'value':taxons }
        ],
        dump=True),

        imp.Source(
            'uniprot-fasta', 'fasta',
            [
                { 'name':'src.data.dir',            'location':uniprotDataPath },
                { 'name':'fasta.taxonId',           'value':ds.getTaxonsAsString() },
                { 'name':'fasta.className',         'value':'org.intermine.model.bio.Protein' },
                { 'name':'fasta.classAttribute',    'value':'primaryAccession' },
                { 'name':'fasta.dataSetTitle',      'value':'UniProt data set' },
                { 'name':'fasta.dataSourceName',    'value':'UniProt' },
                { 'name':'fasta.includes',          'value':'uniprot_sprot_varsplic.fasta' },
                { 'name':'fasta.sequenceType',      'value':'protein' },
                { 'name':'fasta.loaderClassName',   'value':'org.intermine.bio.dataconversion.UniProtFastaLoaderTask' }
            ],
        dump=True),

        imp.Source(
            'uniprot-keywords', 'uniprot-keywords',
            [
                { 'name':'src.data.dir',            'location':uniprotDataPath },
                { 'name':'src.data.dir.includes',   'value':'keywlist.xml' }
            ],
        dump=True)
    ])